### ISSUE TITLE
Transmission won't start - seeing "initgroups: Operation not permitted"

### ISSUE THREAD

**Issue by nicbou** [haugene/docker-transmission-openvpn#1561]
After years without issues, the 1.13 version of this image stopped working, so I had to update. This solved the PIA connection issues, but now transmission won't start.

https://github.com/nicbou/homeserver/blob/master/docker-compose.yml

**Logs**

```
torrents_1         | Starting container with revision: 42d00652323b7b7fbee1689a54219599f4795a04
torrents_1         | Creating TUN device /dev/net/tun
torrents_1         | mknod: /dev/net/tun: File exists
torrents_1         | Using OpenVPN provider: PIA
torrents_1         | Provider PIA has a custom setup script, executing it
torrents_1         | Downloading OpenVPN config bundle openvpn into temporary file /tmp/tmp.JHDALK
torrents_1         | Extract OpenVPN config bundle into PIA directory /etc/openvpn/pia
torrents_1         | Starting OpenVPN using config netherlands.ovpn
torrents_1         | Modifying /etc/openvpn/pia/netherlands.ovpn for best behaviour in this container
torrents_1         | Setting OpenVPN credentials...
torrents_1         | Fri Nov 27 12:52:29 2020 OpenVPN 2.4.9 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Apr 20 2020
torrents_1         | Fri Nov 27 12:52:29 2020 library versions: OpenSSL 1.1.1g  21 Apr 2020, LZO 2.10
torrents_1         | Fri Nov 27 12:52:29 2020 NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
torrents_1         | Fri Nov 27 12:52:29 2020 CRL: loaded 1 CRLs from file [[INLINE]]
torrents_1         | Fri Nov 27 12:52:29 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]143.244.40.232:1198
torrents_1         | Fri Nov 27 12:52:29 2020 UDP link local: (not bound)
torrents_1         | Fri Nov 27 12:52:29 2020 UDP link remote: [AF_INET]143.244.40.232:1198
torrents_1         | Fri Nov 27 12:52:29 2020 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
torrents_1         | Fri Nov 27 12:52:29 2020 [amsterdam419] Peer Connection Initiated with [AF_INET]143.244.40.232:1198
torrents_1         | Fri Nov 27 12:52:31 2020 OpenVPN ROUTE6: OpenVPN needs a gateway parameter for a --route-ipv6 option and no default was specified by either --route-ipv6-gateway or --ifconfig-ipv6 options
torrents_1         | Fri Nov 27 12:52:31 2020 OpenVPN ROUTE: failed to parse/resolve route for host/network: 2000::/3
torrents_1         | Fri Nov 27 12:52:31 2020 TUN/TAP device tun0 opened
torrents_1         | Fri Nov 27 12:52:31 2020 /sbin/ip link set dev tun0 up mtu 1500
torrents_1         | Fri Nov 27 12:52:31 2020 /sbin/ip addr add dev tun0 10.44.112.93/24 broadcast 10.44.112.255
torrents_1         | Fri Nov 27 12:52:31 2020 /etc/openvpn/tunnelUp.sh tun0 1500 1553 10.44.112.93 255.255.255.0 init
torrents_1         | Up script executed with tun0 1500 1553 10.44.112.93 255.255.255.0 init
torrents_1         | Updating TRANSMISSION_BIND_ADDRESS_IPV4 to the ip of tun0 : 10.44.112.93
torrents_1         | Updating Transmission settings.json with values from env variables
torrents_1         | Generating settings.json for Transmission from environment and defaults /etc/transmission/default-settings.json
torrents_1         | Overriding bind-address-ipv4 because TRANSMISSION_BIND_ADDRESS_IPV4 is set to 10.44.112.93
torrents_1         | Overriding download-dir because TRANSMISSION_DOWNLOAD_DIR is set to /triage
torrents_1         | Overriding incomplete-dir because TRANSMISSION_INCOMPLETE_DIR is set to /data/incomplete
torrents_1         | Overriding rpc-port because TRANSMISSION_RPC_PORT is set to 9091
torrents_1         | Overriding speed-limit-up because TRANSMISSION_SPEED_LIMIT_UP is set to 250
torrents_1         | Overriding speed-limit-up-enabled because TRANSMISSION_SPEED_LIMIT_UP_ENABLED is set to true
torrents_1         | Overriding watch-dir because TRANSMISSION_WATCH_DIR is set to /data/watch
torrents_1         | sed'ing True to true
torrents_1         |
torrents_1         | -------------------------------------
torrents_1         | Transmission will run as
torrents_1         | -------------------------------------
torrents_1         | User name:   root
torrents_1         | User uid:    0
torrents_1         | User gid:    0
torrents_1         | -------------------------------------
torrents_1         |
torrents_1         | STARTING TRANSMISSION
torrents_1         | Provider PIA has a script for automatic port forwarding. Will run it now.
torrents_1         | If you want to disable this, set environment variable DISABLE_PORT_UPDATER=yes
torrents_1         | Transmission startup script complete.
torrents_1         | Fri Nov 27 12:52:31 2020 WARNING: OpenVPN was configured to add an IPv6 route over tun0. However, no IPv6 has been configured for this interface, therefore the route installation may fail or may not work as expected.
torrents_1         | Fri Nov 27 12:52:31 2020 Initialization Sequence Completed
torrents_1         | initgroups: Operation not permitted
torrents_1         | port is 29624
torrents_1         | the port has been bound to 29624  Fri Nov 27 12:52:37 UTC 2020
torrents_1         | transmission auth not required
torrents_1         | waiting for transmission to become responsive
torrents_1         | [2020-11-27 12:52:37.477] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:52:47.484] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:52:57.494] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:53:07.504] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:53:17.511] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:53:27.520] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:53:37.530] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:53:47.536] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
torrents_1         | [2020-11-27 12:53:57.545] transmission-remote:  (http://localhost:9091/transmission/rpc/) Couldn't connect to server
```

**Host system:**

Ubuntu 20.04

**Others**

I tried deleting the transmission home directory. That had no effect. The credentials have not changed.

**Comment by haugene** [comment]
Out of curiosity. What happens if you just run the vanilla setup from the README? You don't even have to mount any volumes. And you can bind another port (like 9092) to avoid conflict. So you'll end up something like this:
```
version: '3.3'
services:
    transmission-openvpn:
        cap_add:
            - NET_ADMIN
        environment:
            - OPENVPN_PROVIDER=PIA
            - OPENVPN_CONFIG=france
            - OPENVPN_USERNAME=user
            - OPENVPN_PASSWORD=pass
            - LOCAL_NETWORK=192.168.0.0/16
        logging:
            driver: json-file
            options:
                max-size: 10m
        ports:
            - '9092:9091'
        image: haugene/transmission-openvpn
```

Does it work then? And can you access Transmission web ui?

**Comment by nicbou** [comment]
I will try that next, but it's a considerable hassle. I gave it another go today, and realised the transmission UI gives me the following error message:

```
421: Misdirected Request

Transmission received your request, but the hostname was unrecognized.

To fix this, choose one of the following options:

    - Enable password authentication, then any hostname is allowed.
    - Add the hostname you want to use to the whitelist in settings.

If you're editing settings.json, see the 'rpc-host-whitelist' and 'rpc-host-whitelist-enabled' entries.

This requirement has been added to help prevent DNS Rebinding attacks.
```

Disabling the host whitelist, or setting it to "*" has virtually no effect.

**Comment by nicbou** [comment]
More debug information:

```
bash-5.0# service transmission-daemon status
 * You are attempting to run an openrc service on a
 * system which openrc did not boot.
 * You may be inside a chroot or you may have used
 * another initialization system to boot this system.
 * In this situation, you will get unpredictable results!
 * If you really want to do this, issue the following command:
 * touch /run/openrc/softlevel
bash-5.0# service transmission-daemon restart
 * WARNING: transmission-daemon is already starting
bash-5.0# ps ax
PID   USER     TIME  COMMAND
    1 root      0:00 dumb-init /etc/openvpn/start.sh
    6 root      0:00 openvpn --script-security 2 --up-delay --up /etc/openvpn/t
   45 root      0:00 {update-port.sh} /bin/bash /etc/openvpn/pia/update-port.sh
  293 root      0:00 bash
  437 root      0:00 sleep 10
  438 root      0:00 ps ax
```

**Comment by nicbou** [comment]
When trying to run the /etc/transmission/start.sh script manually, I get a much more meaningful error:

```
bash-5.0# /etc/transmission/start.sh
Up script executed with
ERROR, unable to obtain tunnel address
killing 140
```

**Comment by nicbou** [comment]
I tried running the config you suggested and that container won't start:

```
Starting container with revision: 42d00652323b7b7fbee1689a54219599f4795a04
Creating TUN device /dev/net/tun
mknod: /dev/net/tun: Operation not permitted
chmod: /dev/net/tun: No such file or directory
Using OpenVPN provider: PIA
Provider PIA has a custom setup script, executing it
Downloading OpenVPN config bundle openvpn into temporary file /tmp/tmp.eoniok
Extract OpenVPN config bundle into PIA directory /etc/openvpn/pia
Starting OpenVPN using config france.ovpn
Modifying /etc/openvpn/pia/france.ovpn for best behaviour in this container
Setting OpenVPN credentials...
adding route to local network 192.168.0.0/16 via 172.19.0.1 dev eth0
Mon Nov 30 13:28:48 2020 OpenVPN 2.4.9 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Apr 20 2020
Mon Nov 30 13:28:48 2020 library versions: OpenSSL 1.1.1g  21 Apr 2020, LZO 2.10
Mon Nov 30 13:28:48 2020 NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
Mon Nov 30 13:28:48 2020 CRL: loaded 1 CRLs from file [[INLINE]]
Mon Nov 30 13:28:48 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]143.244.57.231:1198
Mon Nov 30 13:28:48 2020 UDP link local: (not bound)
Mon Nov 30 13:28:48 2020 UDP link remote: [AF_INET]143.244.57.231:1198
Mon Nov 30 13:28:48 2020 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
Mon Nov 30 13:28:48 2020 [paris407] Peer Connection Initiated with [AF_INET]143.244.57.231:1198
Mon Nov 30 13:28:50 2020 OpenVPN ROUTE6: OpenVPN needs a gateway parameter for a --route-ipv6 option and no default was specified by either --route-ipv6-gateway or --ifconfig-ipv6 options
Mon Nov 30 13:28:50 2020 OpenVPN ROUTE: failed to parse/resolve route for host/network: 2000::/3
Mon Nov 30 13:28:50 2020 ERROR: Cannot open TUN/TAP dev /dev/net/tun: No such file or directory (errno=2)
Mon Nov 30 13:28:50 2020 Exiting due to fatal error
```

**Comment by nicbou** [comment]
The plot thickens! The exact same docker configuration runs just fine on my mac. The server starts. If I run it on my Ubuntu server, I still get the result in the original post.

For the record, I deleted one of my earlier comments, because it was wrong.

**Comment by nicbou** [comment]
Seeing that `transmission-daemon` is not running, I tried to start in manually inside the container:

```
exec su --preserve-environment ${RUN_AS} -s /bin/bash -c "/usr/bin/transmission-daemon -g ${TRANSMISSION_HOME} --logfile $LOGFILE"

# RUN_AS = ""
# TRANSMISSION_HOME = "/data/transmission-home"
```

As in the logs above, I get `initgroups: Operation not permitted`. Now I wonder why this line works on a Mac host, but not on a Linux host. If I trim the command, transmission-daemon starts and everything works as it should:

```
/bin/bash -c "/usr/bin/transmission-daemon -g ${TRANSMISSION_HOME}"
```

**Comment by haugene** [comment]
Sorry, lost track here. Did you get any further on this? I just saw another thread where a user said he ran a bunch of VMs and that it worked for all but ubuntu 20.04. But, I'm not sure it's related though. I'm running on Ubuntu 20.04 just fine.

Interesting about the error message though. Especially because you're running as root anyways.
Is it the same if you just try to echo something?
```
exec su --preserve-environment root -s /bin/bash echo hello
```

https://www.linuxquestions.org/questions/linux-newbie-8/su-initgroups-operation-not-permitted-220647/
> By default, most Linux distros do not require you to be a member of a particular group to su to root, but perhaps you've turned that on somehow?

https://www.man7.org/linux/man-pages/man3/initgroups.3.html
> The initgroups() function initializes the group access list by
       reading the group database /etc/group and using all groups of which
       user is a member.  The additional group group is also added to the
       list.

I'm not always sure where the host system ends and the container begins. I'm not too familiar with the kernel or the operations within where containers make their magic. Containers aren't VMs after all, they're just logically separated from your host.
Could this have to do with some special setup of your root user on your host?

**Comment by nicbou** [comment]
> Could this have to do with some special setup of your root user on your host?

That could be the issue. Ubuntu does not allow root login by default; it needs to be manually activated. Perhaps I messed up during that process.

> Is it the same if you just try to echo something?

Yes, any use of the `su` command replicates the issue.

**Comment by nicbou** [comment]
Okay, **I found the solution**. I fixed it.

After a bit of digging, I found something off. The owner and group of everything in `/bin` and `/etc` are `296608:296608`, instead of `root:root`. There is no such user or group. Other Alpine-based images in my project are not like that. This was the case before and after the rebuild.

**This was caused by the "userns_mode: host" line in my docker-compose file. I removed it.** I now face a different problem, but I will tackle it separately, because it's not related to this issue.
